Simplified BotConfig to use less internal pointers

mangelajo opened the pull request:
yaml.Unmarshal will leave the existing values if not provided
and that's a way to specify the defaults, the default label
for ready to review becomes "", which deactivates the feature
if not explicitly set.

Signed-off-by: Miguel Angel Ajo <majopela@redhat.com>

mkolesnik commented on pkg/config/repoconfig/repoconfig.go:
```diff
@@ -10,19 +10,18 @@ import (
-	LabelApproved *struct {
-		Approvals *int
-		Label     *string
+	LabelApproved struct {
```
No this should be a pointer, otherwise you might've not wanted it but even an empty file will now set defaults..

The idea was to allow to set file with
```yaml
label-approved:
```

And it'll fill the defaults of the label/reviewers.

If you want to simplify we can make those mandatory, but this field should still be optional.

mangelajo committed:
Implement some getters for the config and enable unit tests

This avoid the handling of internal pointers on the consumers
of the module and enables unit tests in CI.

Signed-off-by: Miguel Angel Ajo <majopela@redhat.com>

mangelajo commented:
Draft until I add some unit testing

mkolesnik commented on pkg/config/repoconfig/repoconfig.go:
```diff
@@ -15,14 +15,43 @@ const (
+	return c.LabelApproved != nil
+}
+
+func (c *BotConfig) GetLabelApproved() *LabelApprovedConfig {
```
```suggestion
func (c *BotConfig) LabelApproved() *LabelApprovedConfig {
```

This is more "go"

mangelajo replied:
Oh, I didn't know we could shadow the attribute name with a label name, are we sure this is possible?. Let me try.

mangelajo replied:
```
Step 8/16 : RUN go build -o main pkg/main/main.go
 ---> Running in 7b96866b3934
# github.com/submariner-io/pr-brancher-webhook/pkg/config/repoconfig
pkg/config/repoconfig/repoconfig.go:30:6: type BotConfig has both field and method named LabelApproved
```
We can't but, let me wrap the BotConfig as BotConfigYAML

mkolesnik commented on pkg/config/repoconfig/repoconfig.go:
```diff
@@ -15,14 +15,43 @@ const (
+	return &LabelApprovedConfig{}
+}
+
+func (c *LabelApprovedConfig) GetApprovals() int {
```
```suggestion
func (c *LabelApprovedConfig) Approvals() int {
```

mkolesnik commented on pkg/config/repoconfig/repoconfig.go:
```diff
@@ -15,14 +15,43 @@ const (
+		return c.LabelApproved
+	}
+
+	return &LabelApprovedConfig{}
```
```suggestion
	return nil
```

Safer like this, otherwise you're counting on caller to know to call `HasLabelApproved`

mangelajo replied:
May be we should panic instead, with a nicer explanation.

Otherwise we will end up in panic because of a nil de-reference.

I'll make it panic ;-)

mangelajo replied:
I didn't make it panic in the end, it would be weird to panic if somebody tried to do LabelApproved() == nil or something like that.

mkolesnik replied:
Panic will cause bot to crash not a good idea IMHO :)

mkolesnik commented on pkg/config/repoconfig/repoconfig.go:
```diff
@@ -15,14 +15,43 @@ const (
+	return defaultApprovals
+}
 
+func (c *LabelApprovedConfig) GetLabel() string {
```
```suggestion
func (c *LabelApprovedConfig) Label() string {
```

mkolesnik commented on pkg/config/repoconfig/repoconfig.go:
```diff
@@ -15,14 +15,43 @@ const (
+}
+
+func (c *LabelApprovedConfig) GetApprovals() int {
+	if c.Approvals != nil {
+		return *c.Approvals
+	}
+
+	return defaultApprovals
```
```suggestion
	if c.Approvals == nil {
		return defaultApprovals
	}

	return *c.Approvals
```

Main body should be the "expected path", exceptional paths branch out

mkolesnik commented on pkg/config/repoconfig/repoconfig.go:
```diff
@@ -15,14 +15,43 @@ const (
+}
 
+func (c *LabelApprovedConfig) GetLabel() string {
+	if c.Label != nil {
+		return *c.Label
+	}
+
+	return defaultLabel
```
```suggestion
	if c.Label == nil {
		return defaultLabel
	}

	return *c.Label
```

mangelajo replied:
nitpicking detected ;)

mkolesnik reviewed (changes requested):
Initial review (didn't look at tests yet)

mkolesnik commented on pkg/config/repoconfig/repoconfig_suite_test.go:
```diff
@@ -0,0 +1,89 @@
+			config := []byte("label-approved:\n")
+			botConfig, err := parseConfigYAML(config)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(botConfig.HasLabelApproved()).To(BeFalse())
```
Original intention was to allow not setting anything and the config to have defaults.

If we're changing that already, the simplest implementation would be to allow either no label-approved, or if it's specified then it much have the fields.

mangelajo replied:
I'm not changing this. It's the original behavior already, I'm just adding test.

For the yaml parser if you have:

```
label-approved:
```

It will set that field as nil

mkolesnik commented on pkg/config/repoconfig/yaml.go:
```diff
@@ -0,0 +1,34 @@
+	filename         = ".submarinerbot.yaml"
+)
+
+type BotConfigYAML struct {
+	LabelApproved *LabelApprovedYAML `yaml:"label-approved"`
+}
+
+type LabelApprovedYAML struct {
+	Approvals *int
+	Label     *string
+}
```
I understand why you're doing this, but it seems kind of over-engineered to me.

mangelajo replied:
I think it's an ok compromise to be able to process the pointers before any usage.

mkolesnik commented on pkg/handler/pullrequest/handler.go:
```diff
@@ -72,10 +72,10 @@ func openOrSync(gitRepo *git.Git, pr *github.PullRequestPayload, gh ghclient.GH)
 
 	readyToReviewMsg := ""
-	if config != nil && config.LabelApproved != nil {
+	if config.HasLabelApproved() {
```
```suggestion
	if config.LabelApproved() != nil {
```

Seems simpler to me this way (you can get rid of `HasLabelApproved()`)

mangelajo replied:
Yeah, it doesn't provide a lot of value. The patch is about removing pointers exposed to the user, but in the end we need to return the pointer for this one..

mkolesnik commented on pkg/handler/pullrequestreview.go:
```diff
@@ -33,7 +33,7 @@ func handlePullRequestReview(prr github.PullRequestReviewPayload) error {
 	}
 
-	if config.LabelApproved == nil {
+	if !config.HasLabelApproved() {
```
```suggestion
	if config.LabelApproved() == nil {
```

Also here

mkolesnik reviewed (changes requested):
Some thoughts

mangelajo committed:
Remove HasLabelApproved

Signed-off-by: Miguel Angel Ajo <majopela@redhat.com>

mangelajo commented:
@mkolesnik sent some of the suggested changes as a separate commit, since it's been 11 days since your review, sorry ':D

mangelajo commented on pkg/config/repoconfig/repoconfig_suite_test.go:
```diff
@@ -0,0 +1,86 @@
+			config := []byte("")
+			botConfig, err := parseConfigYAML(config)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(botConfig.LabelApproved()).To(BeNil())
```
There was a comment here which was cleared by github.

I'm not changing behavior with this patch, just documenting / testing the existing behavior. I know it's not what you intended, but it's what the yaml parser is giving:

if the yaml is

```
label-approved:
```

Then the LabelApproved field turns to be nil... :-/

skitt commented:
If the goal is to set defaults, can’t we use the `yaml.Unmarshaler` interface?

mangelajo commented:
> If the goal is to set defaults, can’t we use the `yaml.Unmarshaler` interface?

The goal of this patch was not to add defaults.

If you see the defaults were conditional, only when one of the other fields is set, then it's 
added as defaults.

May be some simpler solution to this, would be to stop using pointer/nullable values,
and then have an "enabled" field which is boolean and false by default.

@skitt @mkolesnik thoughts? :)

We would then have

```
label-approved:
   enabled: true
```


```
label-approved:
   enabled: true
   approvals: 3
```
   

@mkolesnik please note that

```
label-approved:
```

remains as nil for the yaml unmarshaling.

mangelajo commented:
@skitt @mkolesnik see my last comment ^

mangelajo commented:
@skitt I mean, if you guys like more this other approach: https://github.com/submariner-io/submariner-bot/pull/43#issuecomment-874086141

Instead of the existing patch version.

I think it would be simpler, same pattern as in other yaml-based stuff, and we get rid of the pointers (which was our initial problem),  and the extra complexity as well..

mkolesnik commented:
> > If the goal is to set defaults, can’t we use the `yaml.Unmarshaler` interface?
> 
> The goal of this patch was not to add defaults.
> 
> If you see the defaults were conditional, only when one of the other fields is set, then it's
> added as defaults.
> 
> May be some simpler solution to this, would be to stop using pointer/nullable values,
> and then have an "enabled" field which is boolean and false by default.

Not sure what this field gives you, if `label-approved` was specified then we can use that and if it wasn't we can ignore it altogether.

> 
> @skitt @mkolesnik thoughts? :)
> 
> We would then have
> 
> ```
> label-approved:
>    enabled: true
> ```
> 
> ```
> label-approved:
>    enabled: true
>    approvals: 3
> ```
> 
> @mkolesnik please note that
> 
> ```
> label-approved:
> ```
> 
> remains as nil for the yaml unmarshaling.

I think the simplest is drop defaults and just expect the fields to be set if `label-approved` is present (ofc due to limitation empty `label-approved` is same as not specified). Then fields won't be pointers either.

dfarrell07 commented:
Per discussion on Scrub call, closing for now. The bot works, happy to re-evaluate if there's more interest.
